packages/angular/ssr/src/common-engine.ts

```diff
@@ -117,32 +117,34 @@ export class CommonEngine {
       return undefined;
     }
 
-    const pathname = canParseUrl(url) ? new URL(url).pathname : url;
-    // Remove leading forward slash.
-    const pagePath = resolve(publicPath, pathname.substring(1), 'index.html');
-
-    if (pagePath !== resolve(documentFilePath)) {
-      // View path doesn't match with prerender path.
-      const pageIsSSG = this.pageIsSSG.get(pagePath);
-      if (pageIsSSG === undefined) {
-        if (await exists(pagePath)) {
-          const content = await fs.promises.readFile(pagePath, 'utf-8');
-          const isSSG = SSG_MARKER_REGEXP.test(content);
-          this.pageIsSSG.set(pagePath, isSSG);
-
-          if (isSSG) {
-            return content;
-          }
-        } else {
-          this.pageIsSSG.set(pagePath, false);
-        }
-      } else if (pageIsSSG) {
-        // Serve pre-rendered page.
-        return fs.promises.readFile(pagePath, 'utf-8');
-      }
+    const { pathname } = new URL(url, 'resolve://');
+    // Do not use `resolve` here as otherwise it can lead to path traversal vulnerability.
+    // See: https://portswigger.net/web-security/file-path-traversal
+    const pagePath = join(publicPath, pathname, 'index.html');
+
+    if (this.pageIsSSG.get(pagePath)) {
+      // Serve pre-rendered page.
+      return fs.promises.readFile(pagePath, 'utf-8');
+    }
+
+    if (!pagePath.startsWith(normalize(publicPath))) {
```

Comment:
`pagePath` should also be normalized so that we compare two normalized paths (otherwise, there might be false negatives if symbolic links are involved).

Reply:
`pagePath` is already normalized as it is constructed using `join` which always return a normalized path

Reply:
You're right, sorry, I missed that part of `join`. So LGTM 😀